File: app/src/main/java/bapspatil/flickoff/MovieRecyclerViewAdapter.java
package bapspatil.flickoff;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;


public class MovieRecyclerViewAdapter extends RecyclerView.Adapter<MovieRecyclerViewAdapter.MovieViewHolder> {

    private ArrayList<Movie> mMoviesArrayList;
    private Context mContext;
    private ItemClickListener mClickListener;

    public interface ItemClickListener {
        void onItemClick(int position, ImageView posterImageView);
    }

    public MovieRecyclerViewAdapter(Context context, ArrayList<Movie> movieArrayList, ItemClickListener itemClickListener) {
        this.mContext = context;
        this.mMoviesArrayList = movieArrayList;
        this.mClickListener = itemClickListener;
    }

    @Override
    public MovieViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.rv_movie_item, viewGroup, false);
        return new MovieViewHolder(view);
    }

    @Override
    public void onBindViewHolder(MovieViewHolder holder, int position) {
        Movie theMovie = mMoviesArrayList.get(position);
        Glide.with(mContext)
                .load(theMovie.getPosterPath())
                .placeholder(R.drawable.tmdb_placeholder)
                .error(R.drawable.tmdb_placeholder)
                .fallback(R.drawable.tmdb_placeholder)
                .centerCrop()
                .into(holder.mPosterImageView);
    }

    @Override
    public int getItemCount() {
        if (mMoviesArrayList == null) return 0;
        else return mMoviesArrayList.size();
    }

    public class MovieViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ImageView mPosterImageView;

        MovieViewHolder(View itemView) {
            super(itemView);
            mPosterImageView = (ImageView) itemView.findViewById(R.id.poster_image_view);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (mClickListener != null)
                mClickListener.onItemClick(getAdapterPosition(), mPosterImageView);
        }
    }
}

File: app/src/main/java/bapspatil/flickoff/ui/main/MovieRecyclerViewAdapter.kt
package bapspatil.flickoff.ui.main

import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityOptionsCompat
import androidx.core.content.ContextCompat.startActivity
import androidx.recyclerview.widget.RecyclerView
import bapspatil.flickoff.databinding.ItemMovieBinding
import bapspatil.flickoff.model.Movie
import bapspatil.flickoff.ui.details.DetailsActivity
import bapspatil.flickoff.ui.base.BaseViewHolder

class MovieRecyclerViewAdapter(private val mMoviesArrayList: ArrayList<Movie>?) : RecyclerView.Adapter<BaseViewHolder>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): MovieViewHolder {
        val itemMovieBinding = ItemMovieBinding.inflate(LayoutInflater.from(viewGroup.context), viewGroup, false)
        return MovieViewHolder(itemMovieBinding)
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.onBind(position)
    }

    override fun getItemCount(): Int {
        return mMoviesArrayList?.size ?: 0
    }

    inner class MovieViewHolder(itemMovieBinding: ItemMovieBinding) : BaseViewHolder(itemMovieBinding.root), MovieItemViewModel.MovieItemViewModelListener {
        private var mItemMovieBinding: ItemMovieBinding = itemMovieBinding
        private lateinit var mMovieItemViewModel: MovieItemViewModel

        override fun onItemClick(movie: Movie, imageView: ImageView) {
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(mItemMovieBinding.root.context as AppCompatActivity, imageView, "posterTransition")
            val intent = Intent(mItemMovieBinding.root.context, DetailsActivity::class.java)
            intent.putExtra("movie", movie)
            startActivity(mItemMovieBinding.root.context, intent, options.toBundle())
        }

        override fun onBind(position: Int) {
            val theMovie = mMoviesArrayList!![position]
            mMovieItemViewModel = MovieItemViewModel(theMovie, this)
            mItemMovieBinding.viewModel = mMovieItemViewModel
            mItemMovieBinding.executePendingBindings()
        }
    }

    fun clearItems() {
        mMoviesArrayList?.clear()
    }

    fun addItems(moviesList: ArrayList<Movie>) {
        mMoviesArrayList?.addAll(moviesList)
        notifyDataSetChanged()
    }
}
